Assets/Scripts/SS3D/Systems/Inventory/Items/Item.cs

```diff
@@ -285,29 +280,59 @@ public void SetContainer(Container newContainer)
             _container = newContainer;
         }
 
-        // TODO: Improve this
-        // we have this to generate icons at start, I do not know how bad it is for performance
-        // if you know anything about it, tell us
+        // Generate preview of the same object, but without stored items.
         [ServerOrClient]
-        public Sprite GenerateNewIcon()
+        public Sprite GenerateIcon()
         {
             RuntimePreviewGenerator.BackgroundColor = new Color(0, 0, 0, 0);
             RuntimePreviewGenerator.OrthographicMode = true;
-
+            // Find stored items
+            AttachedContainer container = GetComponent<AttachedContainer>();
```

Comment:
this part is buggy for items with no containers, and probably for items with multiple containers too. You can safely assume the item script will be at the top of the item game object hierarchy so you can use GetComponentsInChildren to get all attached containers and iterate through that.

Reply:
GetComponentsInChildren will get containers from  contained items

Reply:
Ho right silly me. You can use the containerInteractive reference then, or maybe use GetComponentsInChildren and keep only those that don't have an item script different from the item we want to show up in the hierarchy, that should safely eliminate containers from contained items.

Either way, your changes should work for multiple and no containers.